### DISCUSSION TITLE
FlexRIC  app_kpm_moni for other metrics besides RSRP

### DISCUSSION THREAD

**Discussion by fllay** [srsran/srsRAN_Project#278] [General Help] [answered]
Hi
I am new to RIC programming and it is a bit difficult to find any document about how to code RIC. I like to study how it works and I try to start with simple example that is `xapp_kpm_moni`. From the document there are three metrics are exposed (i.e. “CQI”, “RSRP”, “ RSRQ”) and from `kpm_sm_ric.c`, there is a code snippet

```
  MeasInfo_t dummy_MeasInfo = { 
    .meas_type = KPM_V2_MEASUREMENT_TYPE_NAME, 
    .measName = {.buf = (unsigned char *)"RSRP", .len = 4},
    .labelInfo = &dummylabelInfo, 
    .labelInfo_len = 1
  };
```

this is the only place I can find RSRP metric. If I want to monitor, for example RSRQ, what code should I looking for? I try to change `RSRP` to  `RSRQ` but It still shows RSRP measurement or it is hard code somewhere?

**Comment by pgawlowicz** [comment, marked as answer]
Hi, those are only dummy metric that we used as examples. Currently, we are updating our tutorial, and we will show, how to get real metrics from a gnb over the E2 interface.

**Reply by fllay** [reply]
Thank you. May I ask that will the tutorial include RAN control example?

**Reply by pgawlowicz** [reply]
We do have a couple of RAN control functions implemented. However, we need to find a RIC controller that will allow showing it. It seems that the FlexRIC version we use now does not support it.  I will check the recent development of FlexRIC and see if we can switch to a new version (branch) and show a basic example of RAN control.

**Reply by rolandmartinson7890** [reply]
Since the interfaces are standardized, is it possible to connect a totally different RIC, e.g VMWare RIC (https://telco.vmware.com/content/dam/digitalmarketing/vmware/en/pdf/docs/vmware-ran-intelligent-controller-datasheet.pdf) using apt IP addressing to srsRAN's gNB?

**Reply by pgawlowicz** [reply]
We are doing our best to provide interoperability with standard-compliant 3rd party solutions. However, we are mostly testing with open-source RICs. Therefore, we would be grateful if you could provide us with your feedback on your experience when connecting our E2 agent commercial RIC solutions.

**Reply by pgawlowicz** [reply]
@fllay Please check our updated [tutorial](https://docs.srsran.com/projects/project/en/latest/tutorials/source/flexric/source/index.html), where we use the `br-flexric` branch that allows specifying metric names using a config file that is passed to an xApp.

**Reply by fllay** [reply]
Thank you. It looks a lot easier since it uses configuration file to choose rpm.

**Reply by fllay** [reply]
Unfortunately, I got `Segmentation Fault` from `gnb`.  I am sure that `gnb` is latest version and `flexric` is in `br-flexric` branch. I ran commands in the following order

1. `5gc`
2. `./nearRT-RIC `
3. `gnb`
4. `./xapp_oran_moni -c xapp_mon_e2sm_kpm.conf `

It seems that `gnb` successfuly connects to `nearRY-RIC`. The UE (Samsung Galaxy A33
) can connect to `gnb` and can go to the internet. However, when I ran `./xapp_oran_moni -c xapp_mon_e2sm_kpm.conf `, `gnb` got `Segmentation Fault`.  The output from `./xapp_oran_moni -c xapp_mon_e2sm_kpm.conf `  is


```
pi@pi-IdeaPad:~/src/flexric/build/examples/xApp/c/monitor$ ./xapp_oran_moni -c xapp_mon_e2sm_kpm.conf 
Setting the config -c file to xapp_mon_e2sm_kpm.conf
[LibConf]: loading service models from SM_DIR: /usr/local/lib/flexric/
[LibConf]: reading configuration for xApp
[LibConf]: NearRT_RIC IP: 127.0.0.1
[LibConf]: E42_Port Port: 36422
[LibConf]: Sub_ORAN_SM Name: KPM, Time: 1000
[LibConf]: format 1, RAN type ngran_gNB, actions = DRB.UEThpDl 
[xAap]: Initializing ... 
[xApp]: nearRT-RIC IP Address = 127.0.0.1, PORT = 36422
[E2-AGENT]: Opening plugin from path = /usr/local/lib/flexric/libkpm_sm.so 
[E2-AGENT]: Opening plugin from path = /usr/local/lib/flexric/libtc_sm.so 
[E2-AGENT]: Opening plugin from path = /usr/local/lib/flexric/librc_sm.so 
[E2-AGENT]: Opening plugin from path = /usr/local/lib/flexric/librlc_sm.so 
[E2-AGENT]: Opening plugin from path = /usr/local/lib/flexric/libslice_sm.so 
[E2-AGENT]: Opening plugin from path = /usr/local/lib/flexric/libpdcp_sm.so 
[E2-AGENT]: Opening plugin from path = /usr/local/lib/flexric/libmac_sm.so 
[E2-AGENT]: Opening plugin from path = /usr/local/lib/flexric/libgtp_sm.so 
[NEAR-RIC]: Loading SM ID = 2 with def = ORAN-E2SM-KPM 
[NEAR-RIC]: Loading SM ID = 146 with def = TC_STATS_V0 
[NEAR-RIC]: Loading SM ID = 3 with def = ORAN-E2SM-RC 
[NEAR-RIC]: Loading SM ID = 143 with def = RLC_STATS_V0 
[NEAR-RIC]: Loading SM ID = 145 with def = SLICE_STATS_V0 
[NEAR-RIC]: Loading SM ID = 144 with def = PDCP_STATS_V0 
[NEAR-RIC]: Loading SM ID = 142 with def = MAC_STATS_V0 
[NEAR-RIC]: Loading SM ID = 148 with def = GTP_STATS_V0 
[xApp]: E42 SETUP-REQUEST sent
adding event fd = 5 ev-> 5 
[xApp]: E42 SETUP-RESPONSE received
[xApp]: xApp ID = 9 
Registered E2 Nodes = 1 
Pending event size before remove = 1 
Registered node 0 ran func id = 2 
 [xApp]: reporting period = 1000 [ms]
Generated of req_id = 1 
E42_RIC_SUBSCRIPTION_REQUEST 31 
adding event fd = 5 ev-> 6 
[xApp]: RIC SUBSCRIPTION REQUEST sent
[xApp]: SUBSCRIPTION RESPONSE received
Pending event size before remove = 1 
[xApp]: Successfully SUBSCRIBED to ran function = 2 
[xApp]: E42 UPDATE-E2-NODE received
[xApp]: xApp ID = 9 
[NEAR-RIC]: Removing E2 Node MCC 1 MNC 1 NB_ID 411 
/////// detect E2 nodes len update, new len = 0, old len = 1 ///////
```

Any idea what went wrong?

**Reply by pgawlowicz** [reply]
@fllay does it run correctly with the zmq-based setup?

**Reply by yc541** [reply]
@pgawlowicz I have the same error with a OnePlus Nord. The zmq works with most supported KPMs but not DRB.PacketSuccessRateUlgNBUu. Once this KPM is included in the xapp config the gnb (zmq) somehow freezes. With the USRP setup, the RIC subscription is handled correctly but immediately followed with segmentation fault. I have attached my gnb logs and gnb config file (USRP).

[SRS.zip](https://github.com/srsran/srsRAN_Project/files/14084193/SRS.zip)

**Reply by mas-oliveira** [reply]
Yes I wasn't able to collect DRB.PacketSuccessRateUlgNBUu too. 
Someone successfully add the measures names in KPM Format 1 messages?

**Reply by fllay** [reply]
@pgawlowicz I did not try ZMQ but @yc541 mention that he had tried ZMQ without any problem. I have never try ZMQ before but with previous release, I did not get any segmentation fault on gnb.

**Reply by yc541** [reply]
@fllay I just checked the gNB config, do not forget to include these lines. The segmentation fault goes away when these are added

metrics: 
  rlc_json_enable: 1                # Enable RLC metrics reporting (need to deliver measurements to E2SM_KPM service model)
  rlc_report_period: 1000           # Set reporting period to 1s

**Reply by fllay** [reply]
@yc541  Thank you very much. It works perfectly now.

**Reply by ankuramistry** [reply]
Hi @fllay and @yc541 

I have the same segmentation fault error and i tried adding the metrics as above for rlc but i am getting an error as below when running the gnb file. Any idea on what could be wrong?

Could not convert: --rlc_report_period = 1000

**Reply by robertfalkenberg** [reply]
Please ensure there is no [invisible character](https://www.editpad.org/tool/invisible-character) or similar after the 1000. For example re-type the whole line manually.

**Reply by ankuramistry** [reply]
@robertfalkenberg Thank you ! It worked.
